### PR TITLE
add example of DDIM implementation

### PR DESCRIPTION
**Pull request by yng87**
I created example implementation of denoising diffusion model.
The files are copied from [yng87/DDIM.jl](https://github.com/yng87/DDIM.jl)

c.f. https://discourse.julialang.org/t/lux-jl-implementation-of-denoising-diffusion-model/90275

### CODE REVIEW

**Comment by avik-pal** [comment]
Can you run the JuliaFormatter using the config in Lux? See https://lux.csail.mit.edu/dev/devdocs/style_guide/#style-guide. I will do a more detailed review later.

File: examples/DDIM/model.jl [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    @assert size(angular_speeds) == (1, 1, div(embedding_dims, 2), 1)
+
+    embeddings = cat(sin.(angular_speeds .* x), cos.(angular_speeds .* x); dims=3)
+    @assert size(embeddings) == (1, 1, embedding_dims, size(x, 4))
+
+    return embeddings
+end
+
+# activations
+function sigmoid(x)
```
Comment (avik-pal) [inline examples/DDIM/model.jl:37]: If you do `using NNlib` you don't need to define `sigmoid` and `silu`. These are not numerically stable so it is always recommended to use `NNlib`

File: examples/DDIM/model.jl [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+# Basic building block of UNet
+function residual_block(in_channels::Int, out_channels::Int)
+    if in_channels == out_channels
+        first_layer = NoOpLayer()
+    else
+        first_layer = Conv((3, 3), in_channels => out_channels; pad=SamePad())
+    end
+
+    return Lux.Chain(first_layer,
+                     Lux.SkipConnection(Lux.Chain(BatchNorm(out_channels; affine=false,
```
Comment (avik-pal) [inline examples/DDIM/model.jl:54]: Remove `Lux.` all these are exported

File: examples/DDIM/model.jl [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+function (db::DownBlock)(x::AbstractArray{T, 4}, ps,
+                         st::NamedTuple) where {T <: AbstractFloat}
+    skips = () # accumulate intermediate outputs
+    for i in 1:length(db.residual_blocks)
+        layer_name = Symbol(:layer_, i)
+        x, new_st = db.residual_blocks[i](x, ps.residual_blocks[layer_name],
+                                          st.residual_blocks[layer_name])
+        # Don't use push! on vector because it invokes Zygote error
+        skips = (skips..., x)
+        Lux.@set! st.residual_blocks[layer_name] = new_st
```
Comment (avik-pal) [inline examples/DDIM/model.jl:92]: `using Setfield`. This will break if `Lux` drops dependency on `Setfield` (which is not a breaking change)

File: examples/DDIM/model.jl [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                                                  WrappedFunction(Base.Fix1(broadcast,
+                                                                            silu)),
+                                                  Conv((3, 3), out_channels => out_channels;
+                                                       stride=1, pad=(1, 1))), +))
+end
+
+# Downsampling block of UNet
+# It narrows height and width while increasing channels.
+struct DownBlock <: Lux.AbstractExplicitContainerLayer{(:residual_blocks, :maxpool)}
+    residual_blocks::Lux.Chain
```
Comment (avik-pal) [inline examples/DDIM/model.jl:67]: Too tight a type requirement. If it is a `AbstractExplicitLayer` it should work.

File: examples/DDIM/model.jl [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        Lux.@set! st.residual_blocks[layer_name] = new_st
+    end
+    x, _ = db.maxpool(x, ps.maxpool, st.maxpool)
+    return (x, skips), st
+end
+
+# Upsampling block of UNet
+# It doubles height and width while decreasing channels.
+struct UpBlock <: Lux.AbstractExplicitContainerLayer{(:residual_blocks, :upsample)}
+    residual_blocks::Lux.Chain
```
Comment (avik-pal) [inline examples/DDIM/model.jl:101]: Too tight a type requirement. If it is an `AbstractExplicitLayer` it should work.

File: examples/DDIM/model.jl [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    reverse!(channels)
+    up_blocks = [UpBlock(channels[i], channels[i + 1], block_depth)
+                 for i in 1:(length(channels) - 1)]
+    up_blocks = Lux.Chain(up_blocks...)
+
+    return UNet(upsample, conv_in, conv_out, down_blocks, residual_blocks, up_blocks,
+                noise_embedding)
+end
+
+function (unet::UNet)(x::Tuple{AbstractArray{T, N}, AbstractArray{T, N}}, ps,
```
Comment (avik-pal) [inline examples/DDIM/model.jl:178]: Set `N` to 4. Then you can avoid the assertion checks at runtime.

File: examples/DDIM/model.jl [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    for i in 1:length(unet.down_blocks)
+        layer_name = Symbol(:layer_, i)
+        (x, skips), new_st = unet.down_blocks[i](x, ps.down_blocks[layer_name],
+                                                 st.down_blocks[layer_name])
+        Lux.@set! st.down_blocks[layer_name] = new_st
+        skips_at_each_stage = (skips_at_each_stage..., skips)
+    end
+
+    x, new_st = unet.residual_blocks(x, ps.residual_blocks, st.residual_blocks)
+    Lux.@set! st.residual_blocks = new_st
```
Comment (avik-pal) [inline examples/DDIM/model.jl:211]: Setfield

File: examples/DDIM/model.jl [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+    x, new_st = unet.conv_out(x, ps.conv_out, st.conv_out)
+    Lux.@set! st.conv_out = new_st
+
+    return x, st
+end
+
+# Define DDIM model
+# This generates noise, adds it to images and calls UNet on it to denoise.
+struct DenoisingDiffusionImplicitModel <:
```
Comment (avik-pal) [inline examples/DDIM/model.jl:228]: having abstract types is not ideal. Make it `DDIM{T <: AbstractFloat}` and `(min/max)_signal_rate::T`. Ideally, all the inner fields should be typed, but that often obfuscates the error messages, so it is fine to avoid those.

File: examples/DDIM/train.jl [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+    println("Preparing DDIM.")
+    ddim = DenoisingDiffusionImplicitModel((image_size, image_size); channels=channels,
+                                           block_depth=block_depth, min_freq=min_freq,
+                                           max_freq=max_freq, embedding_dims=embedding_dims,
+                                           min_signal_rate=min_signal_rate,
+                                           max_signal_rate=max_signal_rate)
+    ps, st = Lux.setup(rng, ddim) .|> gpu
+
+    println("Set optimizer.")
+    opt = Optimisers.AdamW(learning_rate, (9.0f-1, 9.99f-1), weight_decay)
```
Comment (avik-pal) [inline examples/DDIM/train.jl:148]: I think `AdamW` is exported?

**Review by avik-pal — CHANGES_REQUESTED** [review]

File: examples/DDIM/model.jl (hunk 1 above)
Comment (yng87) [reply examples/DDIM/model.jl]: Thanks. I used NNlib

File: examples/DDIM/model.jl (hunk 4 above)
Comment (yng87) [reply examples/DDIM/model.jl]: I modified `Chain` to `AbstractExplicitLayer`

File: examples/DDIM/model.jl (hunk 8 above)
Comment (yng87) [reply examples/DDIM/model.jl]: Thanks for the suggestion. I modified it.

**Comment by yng87** [comment]
Thanks for helpful comments! I've addressed all of them.

File: examples/DDIM/train.jl [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+@main function main(; dataset_dir::String, epochs::Int=1, image_size::Int=64,
+                    batchsize::Int=64, learning_rate::Float64=1e-3,
+                    weight_decay::Float64=1e-4, val_diffusion_steps::Int=3,
+                    checkpoint_interval::Int=5, output_dir::String="output/train",
+                    # model hyper params
+                    channels::Vector{Int}=[32, 64, 96, 128], block_depth::Int=2,
+                    min_freq::Float32=1.0f0, max_freq::Float32=1000.0f0,
+                    embedding_dims::Int=32, min_signal_rate::Float32=0.02f0,
+                    max_signal_rate::Float32=0.95f0)
+    rng = Random.Xoshiro()
```
Comment (avik-pal) [inline examples/DDIM/train.jl:120]: Use `MersenneTwister`. `Xoshiro` is not available in julia 1.6

File: examples/DDIM/train.jl [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+    return image
+end
+
+#=
+Training utilities
+=#
+function compute_loss(ddim::DenoisingDiffusionImplicitModel{T}, images::AbstractArray{T, 4},
+                      rng::AbstractRNG, ps, st::NamedTuple) where {T <: AbstractFloat}
+    (noises, images, pred_noises, pred_images), st = ddim((images, rng), ps, st)
+    noise_loss = Statistics.mean(abs.(pred_noises - noises))
```
Comment (avik-pal) [inline examples/DDIM/train.jl:82]:
```suggestion
    noise_loss = mean(abs.(pred_noises - noises))
```

File: examples/DDIM/train.jl [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    image_loss = Statistics.mean(abs.(pred_images - images))
```
Comment (avik-pal) [inline examples/DDIM/train.jl:83]:
```suggestion
    image_loss = mean(abs.(pred_images - images))
```

File: examples/DDIM/train.jl [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+        img = ds.preprocess(img)
+        img = permutedims(channelview(img), (2, 3, 1))
+        if ds.use_cache
+            ds.cache[i] = img
+        end
+        return Float32.(img)
+    end
+end
+
+function center_crop(image::Matrix{RGB{T}}) where {T <: Real}
```
Comment (avik-pal) [inline examples/DDIM/train.jl:50]: Use `DataAugmentation` or `Augmentor` for these. Don't reimplement those functionality

**Review by avik-pal — CHANGES_REQUESTED** [review]

File: examples/DDIM/train.jl (hunk 13 above)
Comment (yng87) [reply examples/DDIM/train.jl]: Thanks. I didn't know those packages.
I've used `Augmentor`

**Review by avik-pal — APPROVED** [review]

File: examples/DDIM/Project.toml [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+NNlib = "872c559c-99b0-510c-b3b7-b6c96a88d5cd"
+Optimisers = "3bd65402-5787-11e9-1adc-39752487f4e2"
+Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
+Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
+ProgressBars = "49802e3a-d2f1-5c88-81d8-b72133a6f568"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+Revise = "295af30f-e4ad-537b-8983-00126c2a3abe"
+Setfield = "efcf1570-3423-57d1-acb7-fd33fddbac46"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
```
Comment (avik-pal) [inline examples/DDIM/Project.toml:19]: Last thing. Let's add compat bounds here and add `examples/DDIM` to https://github.com/avik-pal/Lux.jl/blob/7a3eebb5b4014bca80ca9baf064de013b2eaee8f/.github/workflows/CompatHelper.yml#L30 so that they are automatically updated in the future

File: examples/DDIM/README.md [hunk 15]
```diff
@@ -0,0 +1,43 @@
+# DDIM.jl
```
Comment (avik-pal) [inline examples/DDIM/README.md:1]:
```suggestion
# Denoising Diffusion Implicit Model (DDIM)
```

File: examples/DDIM/Project.toml (hunk 14 above)
Comment (yng87) [reply examples/DDIM/Project.toml:19]: I added `examples/DDIM` in https://github.com/avik-pal/Lux.jl/pull/190/commits/e6c0f1532f714b2fa77ac95f939b1b5e3f73484f

**Comment by yng87** [comment]
Thanks for the review. I've also updated sample images as the random number generator is changed https://github.com/avik-pal/Lux.jl/pull/190/commits/152bd89af5cf137af4f1fb4f8a9b46bde0fee76b

**Comment by avik-pal** [comment]
Thanks for the contribution!!
